### PR TITLE
Avoid panic when stopping the podKiller

### LINKED ISSUE #38857: Dangerous channel using in Kubelet.podKiller
https://github.com/kubernetes/kubernetes/blob/1d3c7ca7581272124b0853a606bb88387a4db9ea/pkg/kubelet/kubelet_pods.go#L756

There is a defer calling in this method:

    defer close(resultCh)

It may crash the goroutine created at line 777, which then will make the whole kubelet crash.

A simple fix
```go
	func (kl *Kubelet) podKiller() {
		killing := sets.NewString()
		resultCh := make(chan types.UID)

		defer close(resultCh)

		for {
			select {
			case podPair, ok := <-kl.podKillingCh:
				if !ok {
					return
				}

				runningPod := podPair.RunningPod
				apiPod := podPair.APIPod

				if killing.Has(string(runningPod.ID)) {
					// The pod is already being killed.
					break
				}
				killing.Insert(string(runningPod.ID))
				go func(apiPod *v1.Pod, runningPod *kubecontainer.Pod, ch chan types.UID) {
					defer func() {
						recover() // avoid this goroutine crashing
					}
					defer func() {
						ch <- runningPod.ID
					}()
					glog.V(2).Infof("Killing unwanted pod %q", runningPod.Name)
					err := kl.killPod(apiPod, runningPod, nil, nil)
					if err != nil {
						glog.Errorf("Failed killing the pod %q: %v", runningPod.Name, err)
					}
				}(apiPod, runningPod, resultCh)

			case podID := <-resultCh:
				killing.Delete(string(podID))
			}
		}
	}
```

Another simple fix
```go
	func (kl *Kubelet) podKiller() {
		killing := sets.NewString()
		resultCh := make(chan types.UID)

		// defer close(resultCh)
		closedCh := make(chan struct{}) // this channel is used to avoid goroutine 777 hanging
		defer close(closedCh)

		for {
			select {
			case podPair, ok := <-kl.podKillingCh:
				if !ok {
					return
				}

				runningPod := podPair.RunningPod
				apiPod := podPair.APIPod

				if killing.Has(string(runningPod.ID)) {
					// The pod is already being killed.
					break
				}
				killing.Insert(string(runningPod.ID))
				go func(apiPod *v1.Pod, runningPod *kubecontainer.Pod, ch chan types.UID) {
					defer func() {
						select {
						case ch <- runningPod.ID:
						case <- closedCh: // avoid this goroutine hanging
						}
					}()
					glog.V(2).Infof("Killing unwanted pod %q", runningPod.Name)
					err := kl.killPod(apiPod, runningPod, nil, nil)
					if err != nil {
						glog.Errorf("Failed killing the pod %q: %v", runningPod.Name, err)
					}
				}(apiPod, runningPod, resultCh)

			case podID := <-resultCh:
				killing.Delete(string(podID))
			}
		}
	}
```
The third fix:
```go
	func (kl *Kubelet) podKiller() {
		exitCh := make(chan struct{})
		defer func() {
			<- exitCh
		}()
		
		go func() {
			killing := sets.NewString()
			resultCh := make(chan types.UID)

			// defer close(resultCh)

			for {
				select {
				case podPair, ok := <-kl.podKillingCh:
					if !ok {
						close(exitCh) // notify podKiller to exit
						return
					}

					runningPod := podPair.RunningPod
					apiPod := podPair.APIPod

					if killing.Has(string(runningPod.ID)) {
						// The pod is already being killed.
						break
					}
					killing.Insert(string(runningPod.ID))
					go func(apiPod *v1.Pod, runningPod *kubecontainer.Pod, ch chan types.UID) {
						defer func() {
							ch <- runningPod.ID
						}()
						glog.V(2).Infof("Killing unwanted pod %q", runningPod.Name)
						err := kl.killPod(apiPod, runningPod, nil, nil)
						if err != nil {
							glog.Errorf("Failed killing the pod %q: %v", runningPod.Name, err)
						}
					}(apiPod, runningPod, resultCh)

				case podID := <-resultCh:
					killing.Delete(string(podID))
				}
			}
		}()
	}
```

### PR DESCRIPTION
**Pull request by fraenkel**
use a mutex instead of a channel

fixes #38857

### CODE REVIEW

**Comment by yujuhong** [comment]
@k8s-bot ok to test

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -799,8 +800,7 @@ func (kl *Kubelet) HandlePodCleanups() error {
 // another goroutine isn't already in action.
 func (kl *Kubelet) podKiller() {
 	killing := sets.NewString()
-	resultCh := make(chan types.UID)
-	defer close(resultCh)
+	lock := sync.Mutex{}
```
Comment (yujuhong) [inline pkg/kubelet/kubelet_pods.go]: nit: add a one-line comment explaining that the mutex is used to guard the mutation of the `killing` set.

**Comment by yifan-gu** [comment]
LGTM

**Comment by yujuhong** [comment]
LGTM pending the nit. Thanks for the fix

**Comment by fraenkel** [comment]
@yujuhong Change is in, failure is a known flake.

**Comment by yujuhong** [comment]
@k8s-bot kops aws e2e test this.

**Pull request merged**